src/main/host/thread.rs

```diff
@@ -56,54 +72,78 @@ impl Thread {
                 ),
                 x => panic!("Bad number of syscall args {}", x),
             }
-        };
-        syscall::raw_return_value_to_result(raw_res)
+        }
+    }
+
+    /// Have the plugin thread natively execute the given syscall.
+    fn native_syscall(&self, n: i64, args: &[SysCallReg]) -> nix::Result<SysCallReg> {
+        syscall::raw_return_value_to_result(self.native_syscall_raw(n, args))
     }
 
     pub fn process_id(&self) -> ProcessId {
-        // Safety: self.cthread initialized in CThread::new.
-        ProcessId::try_from(unsafe { c::thread_getProcessId(self.cthread()) }).unwrap()
+        self.process_id
     }
 
     pub fn host_id(&self) -> HostId {
-        // Safety: self.cthread initialized in CThread::new.
-        unsafe { c::thread_getHostId(self.cthread()) }
+        self.host_id
     }
 
-    pub fn system_pid(&self) -> Pid {
-        // Safety: self.cthread initialized in CThread::new.
-        Pid::from_raw(unsafe { c::thread_getNativePid(self.cthread()) })
+    pub fn native_pid(&self) -> Pid {
+        // Safety: Initialized in Self::new
+        Pid::from_raw(unsafe { c::managedthread_getNativePid(self.mthread.ptr()) })
     }
 
-    pub fn system_tid(&self) -> Pid {
-        // Safety: self.cthread initialized in CThread::new.
-        Pid::from_raw(unsafe { c::thread_getNativeTid(self.cthread()) })
+    pub fn native_tid(&self) -> Pid {
+        // Safety: Initialized in Self::new
+        Pid::from_raw(unsafe { c::managedthread_getNativeTid(self.mthread.ptr()) })
     }
 
     pub fn csyscallhandler(&self) -> *mut c::SysCallHandler {
-        unsafe { c::thread_getSysCallHandler(self.cthread()) }
+        self.syscallhandler.ptr()
     }
 
     pub fn id(&self) -> ThreadId {
-        ThreadId(unsafe { c::thread_getID(self.cthread()).try_into().unwrap() })
+        self.id
+    }
+
+    /// Returns whether the given thread is its thread group (aka process) leader.
+    /// Typically this is true for the first thread created in a process.
+    pub fn is_leader(&self) -> bool {
+        self.id == self.process_id.into()
     }
 
     pub fn syscall_condition(&self) -> Option<SysCallConditionRef> {
-        let syscall_condition_ptr = unsafe { c::thread_getSysCallCondition(self.cthread()) };
-        if syscall_condition_ptr.is_null() {
-            return None;
+        // We can't safely use `as_mut` here, since that would construct a mutable reference,
+        // and we can't prove no other reference exists.
+        let c = self.cond.get().ptr();
+        if c.is_null() {
+            None
+        } else {
+            Some(unsafe { SysCallConditionRef::borrow_from_c(c) })
```

Comment:
Reworded.

```
        // We check the for null explicitly here instead of using `as_mut` to
        // construct and match an `Option<&mut c::SysCallCondition>`, since it's
        // difficult to ensure we're not breaking any Rust aliasing rules when
        // constructing a mutable reference.
```